tritium-metrics/src/main/java/com/palantir/tritium/metrics/TaggedMetricRegistry.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.tritium.metrics;
+
+import com.codahale.metrics.Counter;
+import com.codahale.metrics.ExponentiallyDecayingReservoir;
+import com.codahale.metrics.Gauge;
+import com.codahale.metrics.Histogram;
+import com.codahale.metrics.Meter;
+import com.codahale.metrics.Metric;
+import com.codahale.metrics.MetricRegistry;
+import com.codahale.metrics.Timer;
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Supplier;
+
+/**
+ * Similar to {@link MetricRegistry} but allows tagging of {@link Metric}s.
+ */
+public final class TaggedMetricRegistry {
+
+    private static final TaggedMetricRegistry DEFAULT = new TaggedMetricRegistry();
+
+    private final Map<MetricName, Metric> registry = new ConcurrentHashMap<>();
+
+    public TaggedMetricRegistry() {}
+
+    /**
+     * Get the global default {@link TaggedMetricRegistry}.
+     */
+    public TaggedMetricRegistry getDefault() {
+        return DEFAULT;
+    }
+
+    public Counter counter(MetricName metric) {
+        return getOrAdd(metric, Counter.class, Counter::new);
+    }
+
+    // This differs from MetricRegistry and takes the Gauge directly rather than a Supplier<Gauge>
+    public Gauge gauge(MetricName metric, Gauge gauge) {
+        return getOrAdd(metric, Gauge.class, () -> gauge);
+    }
+
+    public Histogram histogram(MetricName metric) {
+        return getOrAdd(metric, Histogram.class, () -> new Histogram(new ExponentiallyDecayingReservoir()));
+    }
+
+    public Meter meter(MetricName metric) {
+        return getOrAdd(metric, Meter.class, Meter::new);
+    }
+
+    public Timer timer(MetricName metric) {
+        return getOrAdd(metric, Timer.class, Timer::new);
+    }
+
+    public Map<MetricName, Metric> getMetrics() {
+        return Collections.unmodifiableMap(registry);
```

Comment:
Is the intention is that this registry will only contain tagged metrics, and regular `MetricRegistry` contains non-tagged metrics, and we push the non-tagged over to tagged? Do we care about collisions of just the name part of metrics between the two?

Reply:
we can keep them totally separate and log both?

Reply:
+1 to keeping them separate and logging both. I don't think name collisions are an issue here since the new metrics should be uniquely identified by their tags, even if they do have a name clash with a metric in the vanilla registry

Reply:
We can log both, just want to think through what collisions mean, mostly downstream from this when they feed into systems for analysis. I'd expect that the new tagged MetricName won't collide in general as they should have cleaner names anyway. I think its probably ok and we can warn on the metric extraction side to resolve and fix any potential collisions.

Reply:
agreed. think this is a problem today anyway, since two separate services can register metrics with the same name and won't get the `MetricRegistry` validation.